Dye color procedure blocks

Goldorion opened the pull request:
I discovered a setter to change the text color of a text field, so I also added a new procedure block for that. I decided to add 2 new procedure blocks to get a Dye color (from a mapping file and a random color) and using a new output/input type, DyeColor, because I noticed DyeColor can be used for other things, like buttons. Also, a mapping file will allow other generators (like my Fabric generator) to edit values if needed.
Both Dye colors blocks do not have the client icon because DyeColor (colors in general) is not a client-side thing only.

Max094Reikeb reviewed (approved):
Sounds good

KlemenDEV commented:
Could you split dye and text field procedures into two PRs so I can review and leave independent comments for both features? Thanks!

Goldorion commented:
> 
> 
> Could you split dye and text field procedures into two PRs so I can review and leave independent comments for both features? Thanks!

Done

KlemenDEV reviewed (changes requested):
You do not specify extension of dye colors on some of the fields

Goldorion commented:
> 
> 
> You do not specify extension of dye colors on some of the fields

Both blocks without the extension don't need it. The getter of a random dye color returns a random color, so the extension is not needed here. The block to set the color of the text in a text field also don't need the extension because it uses the default dye color getter, so the user can also use the random dye color block (or any other dye color block added with a plugin).

KlemenDEV commented:
> > You do not specify extension of dye colors on some of the fields
> 
> Both blocks without the extension don't need it. The getter of a random dye color returns a random color, so the extension is not needed here. The block to set the color of the text in a text field also doesn't need the extension because it uses the default dye color getter, so the user can also use the random dye color block (or any other dye color block added with a plugin).

I see

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/JavaKeywordsMap.java:
```diff
@@ -85,6 +88,7 @@
 		put("Boolean", new String[] { "boolean", "false" });
 		put("String", new String[] { "String", "\"\"" });
 		put("MCItem", new String[] { "ItemStack", "ItemStack.EMPTY" });
+		put("DyeColor", new String[] { "DyeColor", "DyeColor.WHITE"});
```
Don't forget to add this to your new variable system once I merge here :)

KlemenDEV reviewed (changes requested):
> Skipping procedure block with special fields: dye_color

You should add handling for dye fields to unit tests. See https://github.com/MCreator/MCreator/blob/d0d91c5febc4f75a09a1aa27e5ff531f146bbece/src/test/java/net/mcreator/integration/generator/GTProcedureBlocks.java#L123

MysteriousWolf reviewed (changes requested):
I reviewed your PR and it looks okay, however, I would propose some additions. In the current state, colours can be defined as static variables and used only to change the text field colors, which makes it seem like a half-baked feature, so I have a few ideas that could be implemented to improve that:

- Getters (and optionally setters) for item and block colours (maybe even text field/label, but I don't think that would be of much use). For example, you put an ItemStack type in a block which then returns its colour. An example of this would be getting wool/dye colours from items.
- Comparators for the color type (e.g. is color1 == color2)

Afterwards, I'd suggest you open another PR that extends the text field with the option to set their default colour, since you can now do that with procedures.

Goldorion commented:
> 
> 
> I reviewed your PR and it looks okay, however, I would propose some additions. In the current state, colours can be defined as static variables and used only to change the text field colors, which makes it seem like a half-baked feature, so I have a few ideas that could be implemented to improve that:
> 
>     * Getters (and optionally setters) for item and block colours (maybe even text field/label, but I don't think that would be of much use). For example, you put an ItemStack type in a block which then returns its colour. An example of this would be getting wool/dye colours from items.
> 
>     * Comparators for the color type (e.g. is color1 == color2)
> 
> 
> Afterwards, I'd suggest you open another PR that extends the text field with the option to set their default colour, since you can now do that with procedures.

I will wait this PR https://github.com/MCreator/MCreator/pull/1182 to be merged, so I will add it as a new local variable type.

Goldorion commented:
> 
> 
> I reviewed your PR and it looks okay, however, I would propose some additions. In the current state, colours can be defined as static variables and used only to change the text field colors, which makes it seem like a half-baked feature, so I have a few ideas that could be implemented to improve that:
> 
>     * Getters (and optionally setters) for item and block colours (maybe even text field/label, but I don't think that would be of much use). For example, you put an ItemStack type in a block which then returns its colour. An example of this would be getting wool/dye colours from items.
> 
>     * Comparators for the color type (e.g. is color1 == color2)
> 
> 
> Afterwards, I'd suggest you open another PR that extends the text field with the option to set their default colour, since you can now do that with procedures.

I added a new DyeColor local variable, so we have all basic blocks (I also added the compare block). Concerning getting the color of an item or a block, I looked into the code and it can be possible, but we need to cast the block/item to its own class. I don't think it would be a great idea because we would need to add a condition for each block or item to cast them with the right class. If you really want this block, I'm ok with that, but I wanted to let you know this detail.

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/JavaKeywordsMap.java:
```diff
@@ -18,6 +18,9 @@
 
 package net.mcreator.blockly.java;
 
+import net.mcreator.minecraft.ElementUtil;
```
I don't think this is needed

KlemenDEV commented on plugins/mcreator-core/variables/dyecolor.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Variable will need to define set and get inside variable yaml for both generators to work properly

KlemenDEV commented:
@Goldorion consider implementing these too to add some more use cases to these procedure blocks:

> I reviewed your PR and it looks okay, however, I would propose some additions. In the current state, colours can be defined as static variables and used only to change the text field colors, which makes it seem like a half-baked feature, so I have a few ideas that could be implemented to improve that:
> 
> Getters (and optionally setters) for item and block colours (maybe even text field/label, but I don't think that would be of much use). For example, you put an ItemStack type in a block which then returns its colour. An example of this would be getting wool/dye colours from items.
> Comparators for the color type (e.g. is color1 == color2)
> Afterwards, I'd suggest you open another PR that extends the text field with the option to set their default colour, since you can now do that with procedures.

Goldorion commented:
> @Goldorion consider implementing these too to add some more use cases to these procedure blocks:
> 
> > Getters (and optionally setters) for item and block colours (maybe even text field/label, but I don't think that would be of much use). For example, you put an ItemStack type in a block which then returns its colour. An example of this would be getting wool/dye colours from items.
> 
> Right now the only use case of all this is to set color of text field which is a bit small feature set to consider adding whole new variable type and other related procedure blocks

As I already explained, getting the color of an item would require a condition for every supported item and block. The problem also occurs to set the color.
> Concerning getting the color of an item or a block, I looked into the code and it can be possible, but we need to cast the block/item to its own class. I don't think it would be a great idea because we would need to add a condition for each block or item to cast them with the right class. If you really want this block, I'm ok with that, but I wanted to let you know this detail.

I will check if things not using the DyeColor class are possible.

KlemenDEV commented:
> As I already explained, getting the color of an item would require a condition for every supported item and block. The problem also occurs to set the color.

Eg. dyes return dyecolor so you can return color of these or default color for others.
